crates/biome_configuration/src/editorconfig.rs

```diff
@@ -22,40 +21,194 @@ use crate::{
-    // TODO: use serde_path_to_error to emit better parse diagnostics
-    serde_ini::from_str(s).map_err(|err| {
+    EditorConfig::from_str(s).map_err(|err| {
         EditorConfigDiagnostic::ParseFailed(ParseFailedDiagnostic {
-            source: Some(Error::from(IniError::from(err))),
+            source: Some(Error::from(err)),
         })
     })
 }
 
+#[derive(Debug)]
+pub enum EditorConfigErrorKind {
+    StaticStr(&'static str),
+    ParseBoolError(std::str::ParseBoolError),
+    ParseFormatNumberError(ParseFormatNumberError),
+    // A section must end with `]`
+    MissingSectionEnd,
+}
+impl std::fmt::Display for EditorConfigErrorKind {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::StaticStr(s) => f.write_str(s),
+            Self::ParseBoolError(error) => {
+                write!(f, "{}", error)
+            }
+            Self::ParseFormatNumberError(error) => {
+                write!(f, "{}", error)
+            }
+            Self::MissingSectionEnd => f.write_str("a section must be closed with `]`"),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct EditorConfigError {
+    line_number: u32,
+    kind: EditorConfigErrorKind,
+}
+impl biome_diagnostics::Diagnostic for EditorConfigError {
+    fn category(&self) -> Option<&'static biome_diagnostics::Category> {
+        Some(biome_diagnostics::category!("configuration"))
+    }
+
+    fn severity(&self) -> crate::Severity {
+        crate::Severity::Error
+    }
+
+    fn description(&self, fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(fmt, "{}", self)
+    }
+
+    fn message(&self, fmt: &mut biome_console::fmt::Formatter<'_>) -> std::io::Result<()> {
+        fmt.write_markup(biome_console::markup!({ AsConsoleDisplay(&self) }))
+    }
+}
+impl biome_console::fmt::Display for EditorConfigError {
+    fn fmt(&self, fmt: &mut biome_console::fmt::Formatter<'_>) -> std::io::Result<()> {
+        write!(fmt, "{}", self)
+    }
+}
+impl std::fmt::Display for EditorConfigError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "line {}: {}", self.line_number, self.kind)
+    }
+}
+
+struct AsConsoleDisplay<'a, T>(&'a T);
+impl<T: std::fmt::Display> biome_console::fmt::Display for AsConsoleDisplay<'_, T> {
+    fn fmt(&self, fmt: &mut biome_console::fmt::Formatter<'_>) -> std::io::Result<()> {
+        fmt.write_fmt(format_args!("{}", self.0))
+    }
+}
+
 /// Represents a parsed .editorconfig file, containing only the options that are relevant to biome.
-#[derive(Debug, Clone, Deserialize, Default)]
-#[serde(default)]
+#[derive(Debug, Clone, Default)]
 pub struct EditorConfig {
-    #[serde(deserialize_with = "deserialize_bool_from_string")]
-    root: bool,
-    #[serde(flatten)]
-    options: HashMap<String, EditorConfigOptions>,
+    pub root: bool,
+    pub sections: Vec<EditorConfigSection>,
```

Comment:
I replaced the `HashMap` with a `Vec` to preserve the order. The order is important to preserve the user intention in overriding behaviors.